Detectors/AOD/src/AODProducerWorkflowSpec.cxx

```diff
@@ -194,12 +196,88 @@ void AODProducerWorkflowDPL::addToMFTTracksTable(mftTracksCursorType& mftTracksC
+                                                         mftTracksCursorType& mftTracksCursor)
+{
+  const auto& tpcClusRefs = data.getTPCTracksClusterRefs();
+  const auto& tpcClusShMap = data.clusterShMapTPC;
+  const auto& tpcClusAcc = data.getTPCClusters();
+
+  for (int src = GIndex::NSources; src--;) {
+    int start = trackRef.getFirstEntryOfSource(src);
+    int end = start + trackRef.getEntriesOfSource(src);
+    LOG(DEBUG) << "Unassigned tracks: src = " << src << ", start = " << start << ", end = " << end;
+    for (int ti = start; ti < end; ti++) {
+      TrackExtraInfo extraInfoHolder;
+      auto& trackIndex = GIndices[ti];
+      if (GIndex::includesSource(src, mInputSources)) {
+        if (src == GIndex::Source::MFT) { // MFT tracks are treated separately since they are stored in a different table
+          const auto& track = data.getMFTTrack(trackIndex.getIndex());
+          addToMFTTracksTable(mftTracksCursor, track, collisionID);
+        } else {
+          auto contributorsGID = data.getSingleDetectorRefs(trackIndex);
+          const auto& trackPar = data.getTrackParam(trackIndex);
+          if (contributorsGID[GIndex::Source::ITS].isIndexSet()) {
+            const auto& itsOrig = data.getITSTrack(contributorsGID[GIndex::ITS]);
+            extraInfoHolder.itsClusterMap = itsOrig.getPattern();
+          }
+          if (contributorsGID[GIndex::Source::TPC].isIndexSet()) {
+            const auto& tpcOrig = data.getTPCTrack(contributorsGID[GIndex::TPC]);
+            extraInfoHolder.tpcInnerParam = tpcOrig.getP();
+            extraInfoHolder.tpcChi2NCl = tpcOrig.getNClusters() ? tpcOrig.getChi2() / tpcOrig.getNClusters() : 0;
+            extraInfoHolder.tpcSignal = tpcOrig.getdEdx().dEdxTotTPC;
+            uint8_t shared, found, crossed; // fixme: need to switch from these placeholders to something more reasonable
+            countTPCClusters(tpcOrig, tpcClusRefs, tpcClusShMap, tpcClusAcc, shared, found, crossed);
+            extraInfoHolder.tpcNClsFindable = tpcOrig.getNClusters();
+            extraInfoHolder.tpcNClsFindableMinusFound = tpcOrig.getNClusters() - found;
+            extraInfoHolder.tpcNClsFindableMinusCrossedRows = tpcOrig.getNClusters() - crossed;
+            extraInfoHolder.tpcNClsShared = shared;
+          }
+          if (contributorsGID[GIndex::Source::ITSTPCTOF].isIndexSet()) {
+            const auto& tofMatch = data.getTOFMatch(contributorsGID[GIndex::Source::ITSTPCTOF]);
+            extraInfoHolder.tofChi2 = tofMatch.getChi2();
+            const auto& tofInt = tofMatch.getLTIntegralOut();
+            float intLen = tofInt.getL();
+            extraInfoHolder.length = intLen;
+            float mass = o2::constants::physics::MassPionCharged; // default pid = pion
+            float expSig = tofMatch.getSignal();
+            extraInfoHolder.tofSignal = expSig;
+            float expMom = 0.f;
+            if (expSig > 0) {
+              float tof = expSig - interactionTime;
```

Comment:
And I realize now that this will not work for unassigned tracks (there the `interactionTime` is passed as -1), i.e. the `tofExpMom` will be wrong. In fact, even for assigned tracks the real TOF PID evaluation should use not the nominal interaction time (as passed to fillTrackTablesPerCollision) but precise timeZero.
Now, I am wondering why at all we storing the `tofExpMom`? It can be always recalculated from precise time-of-flight and stored L-integral.

Reply:
Hi @shahor02 apologies I missed a bit the discussion because I was traveling. 
That said, I agree with you that if the TOF expected momentum is computed as in here i.e.:
```
              float expBeta = (intLen / tof / cSpeed);
              expMom = mass * expBeta / std::sqrt(1.f - expBeta * expBeta);
```
writing the expected momentum can be avoided entirely as it can be recomputed on the fly from the quantities in the table.
**Nonetheless** the TOF expected momentum is to be computed with the expected time of pions and not from the measured TOF.
This also simplifies a bit the issue with the `interactionTime` as it does not enter in the computation of the expected momentum.
The procedure would be as in https://github.com/alisw/AliPhysics/blob/bb0e58318cd9d7fc90d659a724643b0f2b1c623e/RUN3/AliAnalysisTaskAO2Dconverter.cxx#L1657-L1665 where we use the integrated times for pions to define the expected momentum.

```
      const AliPID::EParticleType tof_pid = AliPID::kPion;
      const float exp_time = TOFResponse.GetExpectedSignal(track, tof_pid);
      if (exp_time > 0) { // Check that the expected time is positive
        // Expected beta for such hypothesis
        const Float_t exp_beta = (track->GetIntegratedLength() / exp_time / cspeed);

        tracks.fTOFExpMom = AliMathBase::TruncateFloatFraction(
          AliPID::ParticleMass(tof_pid) * exp_beta * cspeed /
            TMath::Sqrt(1. - (exp_beta * exp_beta)),
          mTrack1Pt);
      }
```
With the present implementation in the RUN3 AOD Produce I don't think that it will work as intended in the PID response.

Reply:
Thanks @njacazio , indeed, the calculation should be (please cross-check)
````
float expSig = = tofInt.getTOF(o2::track::PID::Pion);
extraInfoHolder.tofSignal = tofMatch.getSignal();  // to be clarified what reference time we use for unassigned tracks
float expMom = 0.f;
if (expSig > 0) {
  float expBeta = (intLen / expSig / cSpeed);
  expMom = mass * expBeta / std::sqrt(1.f - expBeta * expBeta);
}
extraInfoHolder.tofExpMom = expMom;
````
If you agree, @nburmaso, could you change this.

@jgrosseo concerning your comment:
> In principle, each unassigned track appears in the AmbiguousTrack table where it is assigned a BC slice or a collision array.

Could you point on the code you refer to? I am little bit lost, since -1 is provided as a collision ID for unassigned tracks and I don't see where the track times are used at all.

Reply:
Hi @shahor02 yes, this would work better indeed. Thanks a lot!!

Reply:
Hi @shahor02 
I have adjusted the ambiguous tracks table to what we have discussed in this PR:
https://github.com/AliceO2Group/AliceO2/pull/6629
Maybe it makes my point more clear.